pkg/profiler/cpu/cpu.go

```diff
@@ -318,6 +312,73 @@ func (p *CPU) addUnwindTableForProcess(pid int) {
 	}
 }
 
+// listenEvents listens for events from the BPF program and handles them.
+// It also listens for lost events and logs them.
+func (p *CPU) listenEvents(ctx context.Context, eventsChan <-chan []byte, lostChan <-chan uint64, requestUnwindInfoChan chan<- int) {
+	var group singleflight.Group
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case receivedBytes := <-eventsChan:
+			if len(receivedBytes) == 0 {
+				continue
+			}
+
+			payload := binary.LittleEndian.Uint64(receivedBytes)
+			// Get the 4 more significant bytes and convert to int as they are different types.
+			// On x86_64:
+			//	- unsafe.Sizeof(int(0)) = 8
+			//	- unsafe.Sizeof(uint32(0)) = 4
+			pid := int(int32(payload))
+			switch {
+			case payload&ProcessEncountered == ProcessEncountered:
+				// TODO(kakkoyun): Refactor.
+				ch := group.DoChan(strconv.Itoa(pid), func() (interface{}, error) {
+					return nil, p.obtainRequiredInfoForProcess(ctx, pid)
+				})
+				// Temporary solutions create permanent problems.
+				go func() {
+					select {
+					case result := <-ch:
+						if result.Err != nil {
+							level.Error(p.logger).Log("msg", "failed to obtain required info for process", "pid", pid, "err", result.Err)
+						}
+					case <-ctx.Done():
+					}
+				}()
+			case payload&RequestUnwindInformation == RequestUnwindInformation && p.dwarfUnwindingPolling:
+				// See onDemandUnwindInfoBatcher for consumer.
+				requestUnwindInfoChan <- pid
+			case payload&RequestProcessMappings == RequestProcessMappings:
+				// Populate mappings cache.
+				_, _ = p.processMappings.PopulateMappings(pid)
+			// TODO: even if we populate the mappings for the process we just got
+			// a sample for, we aren't done yet. We also need to extract the debug
+			// information so we can symbolize in the server.
+			// If the process has exited by the time we call `ObjectFileForProcess`,
+			// we might not be able to do it.
+			//
+			// We could maybe copy the file to tmpfs and then run something like:
+			//
+			//	```
+			//	maps := p.processMappings.MapsForPID(pid)
+			//	for _, mf := range maps {
+			//		_, _ = p.objFileCache.ObjectFileForProcess(mf.PID, mf.Mapping)
+			//	}
+			//  ```
+			case payload&RequestRefreshProcInfo == RequestRefreshProcInfo:
+				// Refresh mappings and their unwind info if they've changed.
+				//
+				// TODO: update the mappings cache above.
+				p.bpfMaps.refreshProcessInfo(pid)
+			}
+		case lost := <-lostChan:
+			level.Warn(p.logger).Log("msg", "lost events", "count", lost)
```

Comment:
It's ok to log here in the meantime, but let's ensure we swap this with a Prometheus counter to avoid spamming the logs. We just want to know if it's happening a lot or not, but typically it's not a problem as we are continuously running in the host and losing some samples won't be problematic.